File: cli/src/preprocess.rs
//! Rust implementation of shared/prompts/preprocess.md. Must reproduce
//! shared/fixtures/explain/*/expected.txt byte for byte; the TS twin in
//! web/ follows the same spec.

pub struct Caps {
    pub per_file: usize,
    pub total: usize,
}

impl Default for Caps {
    fn default() -> Self {
        Caps {
            per_file: 400,
            total: 4000,
        }
    }
}

pub enum Matcher {
    Name,
    Segment,
    Suffix,
}

pub struct Rule {
    pub kind: String,
    pub matcher: Matcher,
    pub value: String,
}

pub fn default_rules() -> Vec<Rule> {
    parse_rules(include_str!("../../shared/prompts/exclude.txt"))
}

pub fn parse_rules(s: &str) -> Vec<Rule> {
    s.lines()
        .filter(|l| !l.is_empty() && !l.starts_with('#'))
        .filter_map(|l| {
            let mut it = l.split('\t');
            let kind = it.next()?.to_string();
            let matcher = match it.next()? {
                "name" => Matcher::Name,
                "segment" => Matcher::Segment,
                "suffix" => Matcher::Suffix,
                _ => return None,
            };
            let value = it.next()?.to_string();
            Some(Rule {
                kind,
                matcher,
                value,
            })
        })
        .collect()
}

pub fn classify<'a>(path: &str, rules: &'a [Rule]) -> Option<&'a str> {
    let segs: Vec<&str> = path.split('/').collect();
    for r in rules {
        let hit = match r.matcher {
            Matcher::Name => segs.last() == Some(&r.value.as_str()),
            Matcher::Segment => segs.contains(&r.value.as_str()),
            Matcher::Suffix => path.ends_with(&r.value),
        };
        if hit {
            return Some(&r.kind);
        }
    }
    None
}

/// (files, added, deleted) from numstat text; binary `-` columns count 0.
pub fn stats(numstat: &str) -> (usize, u64, u64) {
    let mut n = 0;
    let mut a = 0;
    let mut d = 0;
    for l in numstat.lines().filter(|l| !l.trim().is_empty()) {
        let mut it = l.splitn(3, '\t');
        let add = it.next().unwrap_or("-");
        let del = it.next().unwrap_or("-");
        n += 1;
        a += add.parse::<u64>().unwrap_or(0);
        d += del.parse::<u64>().unwrap_or(0);
    }
    (n, a, d)
}

pub fn preprocess(diff: &str, commits: &str, numstat: &str, caps: &Caps, rules: &[Rule]) -> String {
    let mut out: Vec<String> = Vec::new();

    let commit_lines: Vec<&str> = commits.lines().filter(|l| !l.trim().is_empty()).collect();
    if !commit_lines.is_empty() {
        out.push(format!("commits: {}", commit_lines.len()));
        for c in &commit_lines {
            out.push(format!("- {c}"));
        }
    }

    let (n, a, d) = stats(numstat);
    out.push(format!("files: {n} (+{a} -{d})"));

    // split into file sections
    let mut sections: Vec<Vec<&str>> = Vec::new();
    for line in diff.lines() {
        if line.starts_with("diff --git ") {
            sections.push(vec![line]);
        } else if let Some(cur) = sections.last_mut() {
            cur.push(line);
        }
    }

    let mut kept: Vec<(String, Vec<&str>)> = Vec::new();
    let mut excluded: Vec<(String, String)> = Vec::new();
    for sec in sections {
        let header = sec[0];
        let path = match header.rfind(" b/") {
            Some(i) => header[i + 3..].to_string(),
            None => header.to_string(),
        };
        let binary = sec
            .iter()
            .any(|l| l.starts_with("Binary files ") || l.starts_with("GIT binary patch"));
        let reason = if binary {
            Some("binary")
        } else {
            classify(&path, rules)
        };
        match reason {
            Some(r) => excluded.push((path, r.to_string())),
            None => kept.push((path, sec)),
        }
    }
    if !excluded.is_empty() {
        excluded.sort();
        out.push("excluded:".to_string());
        for (path, reason) in &excluded {
            out.push(format!("- {path} ({reason})"));
        }
    }
    out.push("---".to_string());

    kept.sort_by(|x, y| x.0.cmp(&y.0));
    let mut total = 0usize;
    for (path, sec) in kept {
        let mut lines: Vec<String> = sec.iter().map(|l| l.to_string()).collect();
        if lines.len() > caps.per_file {
            let dropped = lines.len() - caps.per_file;
            lines.truncate(caps.per_file);
            lines.push(format!("... truncated ({dropped} more lines)"));
        }
        if total + lines.len() > caps.total {
            out.push(format!("... omitted {path} (size cap)"));
            total += 1;
            continue;
        }
        total += lines.len();
        out.extend(lines);
    }

    let mut s = out.join("\n");
    s.push('\n');
    s
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs;
    use std::path::Path;

    #[test]
    fn classifies_paths() {
        let rules = default_rules();
        assert_eq!(classify("Cargo.lock", &rules), Some("lockfile"));
        assert_eq!(classify("web/package-lock.json", &rules), Some("lockfile"));
        assert_eq!(classify("a/node_modules/b/c.js", &rules), Some("vendored"));
        assert_eq!(classify("dist/app.min.js", &rules), Some("minified"));
        assert_eq!(classify("dist/app.js.map", &rules), Some("generated"));
        assert_eq!(classify("src/main.rs", &rules), None);
        // "vendor" must match a segment, not a substring
        assert_eq!(classify("src/vendors.rs", &rules), None);
    }

    #[test]
    fn stats_handles_binary_columns() {
        assert_eq!(stats("1\t2\ta.txt\n-\t-\tb.png\n"), (2, 1, 2));
        assert_eq!(stats(""), (0, 0, 0));
    }

    #[test]
    fn reproduces_golden_fixtures() {
        let base = Path::new(env!("CARGO_MANIFEST_DIR")).join("../shared/fixtures/explain");
        let rules = default_rules();
        let mut cases = 0;
        let mut dirs: Vec<_> = fs::read_dir(&base)
            .expect("fixtures dir")
            .filter_map(|e| e.ok())
            .filter(|e| e.path().is_dir())
            .map(|e| e.path())
            .collect();
        dirs.sort();
        for dir in dirs {
            let read = |name: &str| fs::read_to_string(dir.join(name)).unwrap_or_default();
            let mut caps = Caps::default();
            for l in read("params.txt").lines() {
                if let Some((k, v)) = l.split_once('=') {
                    match k {
                        "per_file_cap" => caps.per_file = v.trim().parse().unwrap(),
                        "total_cap" => caps.total = v.trim().parse().unwrap(),
                        _ => {}
                    }
                }
            }
            let got = preprocess(
                &read("input.diff"),
                &read("input.commits"),
                &read("input.numstat"),
                &caps,
                &rules,
            );
            let want = read("expected.txt");
            assert_eq!(got, want, "fixture {} mismatch", dir.display());
            cases += 1;
        }
        assert!(cases >= 4, "expected at least 4 fixtures, found {cases}");
    }
}

File: web/lib/explain/preprocess.ts
// typescript implementation of shared/prompts/preprocess.md. must
// reproduce shared/fixtures/explain/*/expected.txt byte for byte; the
// rust twin lives in cli/src/preprocess.rs.
import { excludeTxt } from "./shared.gen";

export interface Caps {
  perFile: number;
  total: number;
}

export const defaultCaps: Caps = { perFile: 400, total: 4000 };

export type Matcher = "name" | "segment" | "suffix";

export interface Rule {
  kind: string;
  matcher: Matcher;
  value: string;
}

export function parseRules(s: string): Rule[] {
  const rules: Rule[] = [];
  for (const line of s.split("\n")) {
    if (!line || line.startsWith("#")) continue;
    const [kind, matcher, value] = line.split("\t");
    if (!kind || !value) continue;
    if (matcher !== "name" && matcher !== "segment" && matcher !== "suffix") continue;
    rules.push({ kind, matcher, value });
  }
  return rules;
}

export function defaultRules(): Rule[] {
  return parseRules(excludeTxt);
}

export function classify(path: string, rules: Rule[]): string | null {
  const segs = path.split("/");
  for (const r of rules) {
    const hit =
      r.matcher === "name"
        ? segs[segs.length - 1] === r.value
        : r.matcher === "segment"
          ? segs.includes(r.value)
          : path.endsWith(r.value);
    if (hit) return r.kind;
  }
  return null;
}

export interface Stats {
  files: number;
  added: number;
  deleted: number;
}

// binary "-" columns count 0
export function stats(numstat: string): Stats {
  let files = 0;
  let added = 0;
  let deleted = 0;
  for (const line of numstat.split("\n")) {
    if (!line.trim()) continue;
    const [a, d] = line.split("\t");
    files += 1;
    added += /^\d+$/.test(a) ? parseInt(a, 10) : 0;
    deleted += /^\d+$/.test(d) ? parseInt(d, 10) : 0;
  }
  return { files, added, deleted };
}

// byte-order comparison: js "<" compares utf-16 code units, which
// diverges from the spec's byte order for astral characters
const enc = new TextEncoder();
function byteCompare(a: string, b: string): number {
  const ba = enc.encode(a);
  const bb = enc.encode(b);
  const n = Math.min(ba.length, bb.length);
  for (let i = 0; i < n; i++) {
    if (ba[i] !== bb[i]) return ba[i] - bb[i];
  }
  return ba.length - bb.length;
}

export function preprocess(
  diff: string,
  commits: string,
  numstat: string,
  caps: Caps,
  rules: Rule[]
): string {
  const out: string[] = [];

  const commitLines = commits.split("\n").filter((l) => l.trim());
  if (commitLines.length > 0) {
    out.push(`commits: ${commitLines.length}`);
    for (const c of commitLines) out.push(`- ${c}`);
  }

  const { files, added, deleted } = stats(numstat);
  out.push(`files: ${files} (+${added} -${deleted})`);

  // split into file sections; content before the first header is dropped
  const sections: string[][] = [];
  for (const line of diff.split("\n")) {
    if (line.startsWith("diff --git ")) {
      sections.push([line]);
    } else if (sections.length > 0) {
      sections[sections.length - 1].push(line);
    }
  }
  // a trailing newline in the input yields one empty trailing line in the
  // last section; drop it so section line counts match the rust `lines()`
  const last = sections[sections.length - 1];
  if (last && last[last.length - 1] === "") last.pop();

  const kept: Array<{ path: string; sec: string[] }> = [];
  const excluded: Array<{ path: string; reason: string }> = [];
  for (const sec of sections) {
    const header = sec[0];
    const i = header.lastIndexOf(" b/");
    const path = i >= 0 ? header.slice(i + 3) : header;
    const binary = sec.some(
      (l) => l.startsWith("Binary files ") || l.startsWith("GIT binary patch")
    );
    const reason = binary ? "binary" : classify(path, rules);
    if (reason) excluded.push({ path, reason });
    else kept.push({ path, sec });
  }

  if (excluded.length > 0) {
    excluded.sort(
      (x, y) => byteCompare(x.path, y.path) || byteCompare(x.reason, y.reason)
    );
    out.push("excluded:");
    for (const { path, reason } of excluded) out.push(`- ${path} (${reason})`);
  }
  out.push("---");

  kept.sort((x, y) => byteCompare(x.path, y.path));
  let total = 0;
  for (const { path, sec } of kept) {
    let lines = sec;
    if (lines.length > caps.perFile) {
      const dropped = lines.length - caps.perFile;
      lines = lines.slice(0, caps.perFile);
      lines.push(`... truncated (${dropped} more lines)`);
    }
    if (total + lines.length > caps.total) {
      out.push(`... omitted ${path} (size cap)`);
      total += 1;
      continue;
    }
    total += lines.length;
    out.push(...lines);
  }

  return out.join("\n") + "\n";
}
